### PR TITLE
[SPARK-21769] [SQL] Add a table-specific option for always respecting schemas inferred/controlled by Spark SQL

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
For Hive-serde tables, we always respect the schema stored in Hive metastore, because the schema could be altered by the other engines that share the same metastore. Thus, we always trust the metastore-controlled schema for Hive-serde tables when the schemas are different (without considering the nullability and cases). However, in some scenarios, Hive metastore also could INCORRECTLY overwrite the schemas when the serde and Hive metastore built-in serde are different.

The proposed solution is to introduce a table-specific option for such scenarios. For a specific table, users can make Spark always respect Spark-inferred/controlled schema instead of trusting metastore-controlled schema. By default, we trust Hive metastore-controlled schema.

## How was this patch tested?
Added a cross-version test case

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 1]
```diff
@@ -763,6 +763,47 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+        versionSpark.sql(
+          s"""
+             |CREATE TABLE $tableName
+             |ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.avro.AvroSerDe'
+             |WITH SERDEPROPERTIES ('respectSparkSchema' = 'true')
+             |STORED AS
+             |  INPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat'
+             |  OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat'
+             |LOCATION '$location'
+             |TBLPROPERTIES ('avro.schema.literal' = '$avroSchema')
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: For such an example that requires users setting `TBLPROPERTIES`, it sounds like we are unable to use the `CREATE TABLE USING` command. cc @cloud-fan

**Comment by gatorsmile** [comment]
cc @sameeragarwal @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: There was an argument about whether we should add `TBLPROPERTIES`, and we decided to not add it. I'm totally fine to add it if it's necessary.

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SourceOptions.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
+
+/**
+ * Options for the Parquet data source.
```
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SourceOptions.scala]: nit: update docs

**Comment by sameeragarwal** [comment]
LGTM, thanks! Are these table properties documented somewhere?

**Comment by gatorsmile** [comment]
We might need a dedicated section for documenting all the table-specific conf options.

**Comment by sameeragarwal** [comment]
Merging to master, thanks!

**Pull request closed without merging**